clients/roscpp/src/libros/spinner.cpp

```diff
@@ -30,10 +30,90 @@
 #include "ros/callback_queue.h"
 
 #include <boost/thread/thread.hpp>
-#include <boost/thread/recursive_mutex.hpp>
+#include <boost/thread/mutex.hpp>
 
 namespace {
-  boost::recursive_mutex spinmutex;
+
+/** class to monitor running single-threaded spinners.
+ *
+ *  Calling the callbacks of a callback queue _in order_, requires a unique SingleThreadedSpinner
+ *  spinning on the queue. Other threads accessing the callback queue will probably intercept execution order.
+
+ *  To avoid multiple SingleThreadedSpinners (started from different threads) to operate on the same callback queue,
+ *  this class stores a map of all spinned callback queues.
+ *  If the spinner is single threaded, the corresponding thread-id is stored in the map
+ *  and if other threads will try to spin the same queue, an error message is issued.
+ *
+ *  If the spinner is multi-threaded, the stored thread-id is NULL and future SingleThreadedSpinners
+ *  should not spin this queue.
+ */
+struct SpinnerMonitor
+{
+  /* store spinner information per callback queue:
+     Only as unique single-threaded spinner is allowed and we store its thread id.
+     As multiple multi-threaded spinners are allowed in parallel, their number is counted.
+  */
+  struct Entry
+  {
+    Entry(const boost::thread::id &tid) : tid(tid) {}
+
+    boost::thread::id tid; // thread id of single-threaded spinner
+    unsigned int num_multi_threaded; // number of multi-threaded spinners
```

Comment:
I don't agree that `in-order` processing is not important if the programmer supports multi-threading. Those are very different dimensions. Multi-threading allows parallel processing and is useful for long running things but they can still be stacked and sequenced appropriately. 

Using an example of processing a video stream. You can currently trivially subscribe with a multi-threaded spinner and process the frames in the order they come in and have a book keeping process that locks correctly such that the N parallel threads return in the order they were started or based on their timestamp sorting. If there's a older message in one thread hold it until older ones have been popped out of the parallel processing. 

However if you break the in order assumption you no longer can assume that the number of threads are the latest N messages. And if you care about the order, (say you're processing video frames) you now have to have a cache of the latest messages and insert the frames back into a sorted manner, holding them for some duration at which you expect that the ordering will not be out of order. 

To effectively process data which may come in out of order we need to provide enough information for developers to reliably reorder the data. Off the top of my head how many items might be out of order, how long they might need to wait for the older messages. And the really important thing to be able to tell developers is how much extra computational overhead and latency will this requirement to resort the data cost them?

In order arrival is currently something that can be relied upon in the transport, and since our queue is in order I think it's worth the effort to keep that guarentee alive and make sure that the interleaving queues pop data in order as well.